<p><center><img src="./logo_finsurf.png?raw=true" title="FINSURF" /></center></p>

<h2>Introduction</h2>

<p>FINSURF (Functional Identification of Non-coding Sequences Using Random Forests) is a tool designed to analyse lists of sequences variants in the human genome. </p>

<p>It assigns a score to each variant, reflecting its functional importance and therefore its likelihood to disrupt the physiology of its carrier. FINSURF scores Single Nucleotide Variants (SNV), insertions and deletions. Among SNVs, transitions and transversions are treated separately. 
Insertions are characterised by a score given to each base flanking the insertion point. Deletions are characterised by a score at every deleted base. FINSURF can (optionally) use a list of known or suspected disease genes, in order to restrict results to variants overlapping cis-regulatory elements linked to these genes. </p>



<h2>Priorizing user's variant</h2>

<p>For a variant of interest, users can generate a graphical representation of "feature contributions Â», showing the relative contributions of genomic, functional or evolutionary information to its score.</p>
<p> Code is available on the <a href="https://github.com/DyogenIBENS/FINSURF.git">master branch</a> of <a href="https://github.com/DyogenIBENS/FINSURF/">FINSURF github repository</a> </p>

<p>FINSURF is implemented as python3 scripts.</p>


<h2>FINSURF model generation</h2>

<p>The generation of finsurf model and the generation of figures from the manuscript can be done with the jupiter notebooks available on the <a href="https://github.com/DyogenIBENS/FINSURF/tree/finsurf-model-generation">finsurf-model-generation branch</a> of FINSURF github repository</p>


